src/http/h1/decode.rs

```diff
@@ -290,4 +347,75 @@ mod tests {
         assert_eq!(e.kind(), io::ErrorKind::Other);
         assert_eq!(e.description(), "early eof");
     }
+
+    #[test]
+    fn test_read_chunked_single_read() {
+        let content = b"10\r\n1234567890abcdef\r\n0\r\n";
+        let mut mock_buf = io::Cursor::new(content);
+        let mut buf = [0u8; 16];
+        let count = Decoder::chunked().decode(&mut mock_buf, &mut buf).expect("decode");
+        assert_eq!(16, count);
+        let result = String::from_utf8(buf.to_vec()).expect("decode String");
+        assert_eq!("1234567890abcdef", &result);
+    }
+
+    fn read_async(mut decoder: Decoder, content: &[u8], block_at: usize, read_buffer_size: usize) -> String {
+        let content_len = content.len();
+        let mock_buf = io::Cursor::new(content.clone());
+        let mut ins = Async::new(mock_buf, block_at);
+        let mut outs = vec![];
+        loop {
+            let mut buf = vec![0; read_buffer_size];
+            match decoder.decode(&mut ins, buf.as_mut_slice()) {
+                Ok(0) => break,
+                Ok(i) => outs.write(&buf[0..i]).expect("write buffer"),
+                Err(e) => {
+                    if e.kind() != io::ErrorKind::WouldBlock {
+                        break;
+                    }
+                    ins.block_in(content_len); // we only block once
+                    0 as usize
+                }
+            };
+        }
+        String::from_utf8(outs).expect("decode String")
+    }
+
+    fn all_async_cases(content: &str, expected: &str, decoder: Decoder) {
```

Comment:
Pretty clever, though it took me reading over it several times before I understood what you were trying to test specifically.